forte/processors/data_augment/algorithms/skeleton_op.py

```diff
@@ -0,0 +1,765 @@
+        # of insertion, and the length is the length of the inserted string.
+        self._inserted_annos_pos_len: DefaultDict[
+            int, SortedDict[int, int]
+        ] = defaultdict(lambda: SortedDict({}))
+
+        # :attr:`_deleted_annos_id`: {datapack id: Set[annotation tid]}
+        # It records the deleted span ids, mapping from datapack id
+        # to a set of annotation tids appended by :func:`delete`.
+        self._deleted_annos_id: DefaultDict[int, Set[int]] = defaultdict(set)
+
+        # :attr:`_data_pack_map`: {orig pack id: new pack id}
+        # It maintains a mapping from the pack id
+        # of the original pack to the pack id of augmented pack.
+        # It is used when copying the MultiPackLink and MultiPackGroup.
+        self._data_pack_map: Dict[int, int] = {}
+
+        # :attr:`_entry_maps`: {datapack id: Dict{orig tid, new tid}}
+        # It is a map for tracking the annotation ids
+        # before and after the auto align. It maps the
+        # original annotation tid to the new annotation tid.
+        # It is used when copying the Link/Group/MultiPackLink/MultiPackGroup.
+        self._entry_maps: Dict[int, Dict[int, int]] = {}
+
+        self.configs = configs
+
+    def perform_augmentation(self, input_pack: DataPack) -> DataPack:
+        r"""
+        Function to apply the defined augmentation function and
+        instantiating it into a data pack. This data pack is then
+        returned.
+
+        Args:
+            data_pack: The Datapack holding the replaced annotations.
+            replaced_annotations: A SortedList of tuples(span, new string).
+            The text and span of the annotations will be updated
+            with the new string.
+
+        Returns:
+            A new data_pack holds the text after replacement.
+        """
+        self.augment(input_pack)
+        augmented_data_pack = self._apply_augmentations(input_pack)
+        return augmented_data_pack
+
+    def _copy_link_or_group(
+        self,
+        entry: Union[Link, Group],
+        entry_map: Dict[int, int],
+        new_pack: DataPack,
+    ) -> bool:
+        r"""
+        This function copies a Link/Group in the data pack.
+        If the children Link/Group does not exist, it will recursively
+        create the children Link/Group. If the children Annotation
+        does not exist, it will abort and return False.
+        Args:
+            entry: The Link/Group in the original data pack to copy.
+            entry_map: The dictionary mapping original entry to copied entry.
+            new_pack: The new data pack, which is the destination of copy.
+        Returns:
+            A bool value indicating whether the copy happens.
+        """
+
+        # If the entry has been copied, return True.
+        if entry.tid in entry_map:
+            return True
+
+        # The entry should be either Link or Group.
+        is_link: bool = isinstance(entry, Link)
```

Comment:
This didn't check for group, shalle we check for group and raise error otherwise?